### PR TITLE
Support for Bitwise Operations as InfixOperations

### PR DESCRIPTION
**Pull request by sjaveed**
I've added support for bitwise AND, OR, XOR, left shift and right shift as infix operations.  This PR contains the new functionality, tests and documentation updates.  This is a new PR that replaces #405 where specs weren't passing possibly due to mis-merged upstream changes.

#### Examples ####
```ruby
users.where((users[:bitmap] & 16).gt(0)).project(Arel.sql('*')) # => SELECT * FROM "users"  WHERE ("users"."bitmap" & 16) > 0
users.where((users[:bitmap] | 16).gt(0)).project(Arel.sql('*')) # => SELECT * FROM "users"  WHERE ("users"."bitmap" | 16) > 0
users.where((users[:bitmap] ^ 16).gt(0)).project(Arel.sql('*')) # => SELECT * FROM "users"  WHERE ("users"."bitmap" ^ 16) > 0
users.where((users[:bitmap] << 1).gt(0)).project(Arel.sql('*')) # => SELECT * FROM "users"  WHERE ("users"."bitmap" << 1) > 0
users.where((users[:bitmap] >> 1).gt(0)).project(Arel.sql('*')) # => SELECT * FROM "users"  WHERE ("users"."bitmap" >> 1) > 0
```

#### Questions: ####
* Is this useful to anyone else?
* Should I implement this in a different way?
* How do I implement the bitwise NOT operator (~) which is unary?

Thanks!

### CODE REVIEW

**Comment by sjaveed** [comment]
Thoughts @rafaelfranca?

**Comment by rafaelfranca** [comment]
LGTM but it needs a rebase now. Could you do it?

File: lib/arel/nodes/infix_operation.rb [hunk 1]
```diff
@@ -40,5 +40,35 @@ def initialize left, right
...
+        super(:<<, left, right)
+      end
+    end
+
+    class BitwiseShiftRight < InfixOperation
+      def initialize left, right
+        super(:>>, left, right)
+      end
+    end
+
```
Comment (rafaelfranca) [inline lib/arel/nodes/infix_operation.rb]: :scissors: this blank line

File: lib/arel/nodes/unary_operation.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
+module Arel
+  module Nodes
+
```
Comment (rafaelfranca) [inline lib/arel/nodes/unary_operation.rb]: :scissors: this blank line.

Comment (sjaveed) [inline lib/arel/nodes/unary_operation.rb]: @rafaelfranca this blank line follows the pattern in infix_operation.rb.  Did you want me to remove it from there as well or just leave it here?

Comment (rafaelfranca) [inline lib/arel/nodes/unary_operation.rb]: Just leave it here.

**Comment by sjaveed** [comment]
Rebased.  In the failing specs, one is failing to install jruby-head (`rvm install jruby-head --binary`) and the other is failing to execute `gem build arel.gemspec`.  Is there a way to re-run the specs to see if at least the jruby-head one passes?

**Comment by sjaveed** [comment]
Ok, retried it on my fork and jruby-head specs seem to pass just fine.  The ruby-head specs also pass but it fails when trying to run `gem build arel.gemspec`.  Running that same command locally under ruby-head fails with a NameError too.  Thoughts @rafaelfranca 

https://travis-ci.org/sjaveed/arel/builds/100582392

**Comment by rafaelfranca** [comment]
Since you added a new file to the repository the gemspec needs to be
updated. Run rake arel.gemspec and commit the changes.

**Comment by sjaveed** [comment]
Looks like it's still failing.  Locally `gem build arel.gemspec` doesn't work but if I prepend bundle exec to it,  everything's peachy: `bundle exec gem build arel.gemspec`.  Might be this needs a change to the .travis.yml file?

**Comment by rafaelfranca** [comment]
Don't worry about that. I'll fix it. Could you squash your commits?

**Comment by sjaveed** [comment]
Commits squashed as requested :-)

**Comment by chriswoodrich** [comment]
good stuff @sjaveed !

**Comment by sjaveed** [comment]
Thanks @chriswoodrich!

**Pull request merged**